### PR TITLE
feat: All stories skeleton 

### PR DESCRIPTION
**Pull request by hanlinc27**
ezgif com-gif-maker

### CODE REVIEW

File: ui/src/pages/AllStories.tsx [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+import { colors } from "../styles/colors";
+import { Story } from "../types";
+
+declare module "@material-ui/core/styles/withStyles" {
+  interface BaseCSSProperties {
+    flip?: boolean;
+  }
+}
+
+const styles = (theme: Theme) =>
```
Comment (hanlinc27) [inline ui/src/pages/AllStories.tsx:10]: Could be changed to `styled-components `but I pretty much left the styling the same as the example.

File: ui/src/pages/AllStories.tsx [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+    noClick: {
+      cursor: "initial",
+    },
+  });
+
+interface ColumnData {
+  dataKey: string;
+  label: string;
+  numeric?: boolean;
+  toggle?: true;
```
Comment (hanlinc27) [inline ui/src/pages/AllStories.tsx]: I wanted to do something like `   {columns[columnIndex].toggle && (
          <Switch
            checked={checked}
            onChange={handleChange}
            color="primary"
          />
        )}`
but hooks should not be called conditionally.

File: ui/src/pages/AllStories.tsx [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+
+  getRowClassName = ({ index }: Row) => {
+    const { classes, onRowClick } = this.props;
+
+    return clsx(classes.tableRow, classes.flexContainer, {
+      [classes.tableRowHover]: index !== -1 && onRowClick != null,
+    });
+  };
+
+  cellRenderer: TableCellRenderer = ({ cellData, columnIndex }) => {
```
Comment (hanlinc27) [inline ui/src/pages/AllStories.tsx]: Can probably simplify the default cell header and column rendering logic from MUI

File: ui/src/pages/AllStories.tsx [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+          </Table>
+        )}
+      </AutoSizer>
+    );
+  }
+}
+
+const VirtualizedTable = withStyles(styles)(MuiVirtualizedTable);
+
+interface TableData {
```
Comment (hanlinc27) [inline ui/src/pages/AllStories.tsx]: Should I be defining a new interface for this? Or just loading in from `Story` types?

File: ui/src/pages/AllStories.tsx [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+  country: string,
+  is_visible: boolean
+): TableData {
+  return { id, title, city, year, author, country, is_visible };
+}
+
+export const AllStories: React.FC = () => {
+  const { data: allStories, error } = useSWR<Story[]>("/api/stories");
+
+  if (allStories) {
```
Comment (hanlinc27) [inline ui/src/pages/AllStories.tsx:69]: Need an **if statement** here, otherwise we get undefined at [0] errors.

File: ui/src/pages/AllStories.tsx (hunk 2 above)
Comment (hanlinc27) [reply ui/src/pages/AllStories.tsx]: ^ in cell rendering

File: ui/src/pages/AllStories.tsx (hunk 1 above)
Comment (DinuWije) [reply ui/src/pages/AllStories.tsx:10]: I think this is fine considering the demo you posted looks good and we are tight on time to finish the project. Correct me if I'm wrong, but we can still mix styled-components in with createStyles elements right?

File: ui/src/pages/AllStories.tsx (hunk 2 above)
Comment (DinuWije) [reply ui/src/pages/AllStories.tsx]: I may be misinterpreting your comment, but I think it's fine to call hooks conditionally as long as they aren't **instantiated** conditionally. @AbhiPrasad pls take a look at this too in case I'm very wrong lmao.

File: ui/src/pages/AllStories.tsx [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+    const rows: TableData[] = [];
+    for (let i = 0; i < 33; i += 1) {
```
Comment (DinuWije) [inline ui/src/pages/AllStories.tsx]: Is the 33 here in the for statement because we currently have 33 stories? Can we put a variable instead (ie. allStories.length())?

File: ui/src/types/index.tsx [hunk 7]
```diff
@@ -12,7 +12,7 @@ export interface Story extends Model {
   title: string;
   content: string;
   current_city: string;
-  year: number;
+  year: string;
```
Comment (DinuWije) [inline ui/src/types/index.tsx]: what was the reasoning behind switching this to a string?

File: ui/package.json [hunk 8]
```diff
@@ -30,13 +30,14 @@
     "leaflet.markercluster": "^1.4.1",
     "material-ui-dropzone": "^3.5.0",
     "prop-types": "^15.7.2",
-    "react": ">= 16.8",
-    "react-dom": ">= 16.8",
+    "react": "^15.3.0 || ^16.0.0-alpha",
+    "react-dom": "^15.3.0 || ^16.0.0-alpha",
```
Comment (AbhiPrasad) [inline ui/package.json:34]: Why the changes here? We need 16.8 for hooks.

File: ui/src/pages/AllStories.tsx (hunk 2 above)
Comment (AbhiPrasad) [reply ui/src/pages/AllStories.tsx]: Yup according to the rules of hooks you cannot do something like this:

```ts
if (blah) {
  const [state, setter] = React.useState();
}
```

To understand this, you can imagine each useState call being a representing an index in an state array.

So imagine a ReactElement class that represents a React component in the virtual dom.

```
ReactElement {
  state: [state1, state2, state3, ...],
  effects: [...],
  callbacks: [...],
  ref: ...,
  children: [ ... ]
}
```

The first time a component is created, this `state` array is instantiated based on how many times `React.useState()` was called. The `setState` setters from `React.useState()` then rely on the array index to update state. Therefore if you don't call a `React.useState()` because of conditional reasons on a re-render, state/state-setter might reference the wrong index of the state array.

File: ui/src/pages/AllStories.tsx (hunk 1 above)
Comment (AbhiPrasad) [reply ui/src/pages/AllStories.tsx:10]: Yeah let's just escape hatch for now, we can come back to this.

File: ui/src/pages/AllStories.tsx [hunk 9]
```diff
@@ -0,0 +1,288 @@
...
+  Table,
+  TableCellRenderer,
+  TableHeaderProps,
+} from "react-virtualized";
+import useSWR from "swr";
+
+import { colors } from "../styles/colors";
+import { Story } from "../types";
+
+declare module "@material-ui/core/styles/withStyles" {
```
Comment (AbhiPrasad) [inline ui/src/pages/AllStories.tsx]: let's move this into `global.d.ts` plz

File: ui/src/pages/AllStories.tsx [hunk 10]
```diff
@@ -0,0 +1,288 @@
...
+interface MuiVirtualizedTableProps extends WithStyles<typeof styles> {
+  columns: ColumnData[];
+  headerHeight?: number;
+  onRowClick?: () => void;
+  rowCount: number;
+  rowGetter: (row: Row) => TableData;
+  rowHeight?: number;
+}
+
+class MuiVirtualizedTable extends React.PureComponent<
```
Comment (AbhiPrasad) [inline ui/src/pages/AllStories.tsx]: Can we extract this into it's own component (`Table.tsx`) and make them all functional? We aren't using any of the class based lifecycles.

We can just turn the class function into standalone functions.

Comment (AbhiPrasad) [reply ui/src/pages/AllStories.tsx]: it costs nothing to pass down additional function args instead of using `this.props`

File: ui/src/pages/AllStories.tsx [hunk 11]
```diff
@@ -0,0 +1,288 @@
...
+    },
+    noClick: {
+      cursor: "initial",
+    },
+  });
+
+interface ColumnData {
+  dataKey: string;
+  label: string;
+  numeric?: boolean;
```
Comment (AbhiPrasad) [inline ui/src/pages/AllStories.tsx]: we shouldn't mix undefined and boolean types. Let's enforce this to be a boolean and write a data transform if needed.

File: ui/src/pages/AllStories.tsx (hunk 2 above)
Comment (AbhiPrasad) [reply ui/src/pages/AllStories.tsx]: Also toggle should just be a boolean, we can handle taking care of it when we grab the data instead of typing the component weirdly.

File: ui/src/pages/AllStories.tsx [hunk 12]
```diff
@@ -0,0 +1,288 @@
...
+function createData(
+  id: number,
+  title: string,
+  city: string,
+  year: string,
+  author: string,
+  country: string,
+  is_visible: boolean
+): TableData {
+  return { id, title, city, year, author, country, is_visible };
```
Comment (AbhiPrasad) [inline ui/src/pages/AllStories.tsx]: Can we memoize based on id? Aka keep a running hashmap that stores id -> `TableData` object?

File: ui/src/pages/AllStories.tsx (hunk 4 above)
Comment (AbhiPrasad) [reply ui/src/pages/AllStories.tsx]: I think we should prefer just using the Story type.

File: ui/src/pages/AllStories.tsx [hunk 13]
```diff
@@ -0,0 +1,288 @@
...
+}
+
+export const AllStories: React.FC = () => {
+  const { data: allStories, error } = useSWR<Story[]>("/api/stories");
+
+  if (allStories) {
+    const rows: TableData[] = [];
```
Comment (AbhiPrasad) [inline ui/src/pages/AllStories.tsx]: prefer `array.map` here.

So something like:

```
const { data: allStories, error } = useSWR<Story[]>("/api/stories");

if (error) return <div>Error returning stories data!</div>;
if (allStories === undefined) return <div>Loading all stories table..</div>;

const rows = allStories.map((story, i) => createData(i, story.title, ...));
```

File: ui/package.json (hunk 8 above)
Comment (AbhiPrasad) [reply ui/package.json:34]: We should just pin to `16.8` directly

**Review by AbhiPrasad — CHANGES_REQUESTED** [review]
1st pass review

File: ui/src/types/index.tsx (hunk 7 above)
Comment (hanlinc27) [reply ui/src/types/index.tsx]: The endpoint on the backend looks like this: 
image
so I just updated the types.

Comment (Shehryar21) [reply ui/src/types/index.tsx]: So i think its better to keep it int because when stories are returned, year is an int. It is just that in formdata everything is in string format, so even if you send it as an int, it would be received in the backend as a string.

Comment (hanlinc27) [reply ui/src/types/index.tsx]: Changed it back 👍🏻

**Comment by hanlinc27** [comment]
Closing this and opening a new PR with updated changes/different approach.
